### ISSUE TITLE
[ENH] Enable caching

### ISSUE THREAD

**Issue by paoloantinori** [fabric8io/fabric8-installer#80]
I'm looking into adding caching capabilities to save from the need to hit the internet each time you have to re-download docker images.

I have done this for plain docker in past with a Squid based solution, I want to check if it can be turned into a generic solution.

**Comment by paoloantinori** [comment]
This is trickier than I was expecting:

Obviously I have to configure SQUID to cache HTTPS traffic, and with some mangling + importing a self signed certificate and restarting docker daemon it did work.

But accordingly to SQUID logs some of the traffic was HIT while some other was still a MISS:

`TCP_MEM_HIT/200 1719 GET https://auth.docker.io/token? - HIER_NONE/- application/json`

vs.

`TCP_MISS/307 1170 GET https://registry-1.docker.io/v2/sonatype/nexus/blobs/sha256:6b9b106ccb616528c20e74b5e3ed69350259f9a6985b219c576ca54e1aaa4924 - PINNED/54.88.227.214 application/octet-stream`

```
1460965301.926    560 127.0.0.1 NONE/200 0 CONNECT registry-1.docker.io:443 - HIER_DIRECT/54.88.227.214 -
1460965302.131    205 127.0.0.1 TCP_MISS/401 517 GET https://registry-1.docker.io/v2/ - PINNED/54.88.227.214 application/json
1460965302.745    611 127.0.0.1 NONE/200 0 CONNECT auth.docker.io:443 - HIER_DIRECT/54.210.75.104 -
1460965302.746      0 127.0.0.1 TCP_MEM_HIT/200 1719 GET https://auth.docker.io/token? - HIER_NONE/- application/json
1460965303.359    613 127.0.0.1 NONE/200 0 CONNECT registry-1.docker.io:443 - HIER_DIRECT/54.88.227.214 -
1460965303.770    409 127.0.0.1 TCP_MISS/200 16372 GET https://registry-1.docker.io/v2/sonatype/nexus/manifests/latest - PINNED/54.88.227.214 application/vnd.docker.distribution.manifest.v1+prettyjws
1460965304.154    381 127.0.0.1 NONE/200 0 CONNECT registry-1.docker.io:443 - HIER_DIRECT/54.88.227.214 -
1460965304.164    392 127.0.0.1 NONE/200 0 CONNECT registry-1.docker.io:443 - HIER_DIRECT/54.88.227.214 -
1460965304.170    398 127.0.0.1 NONE/200 0 CONNECT registry-1.docker.io:443 - HIER_DIRECT/54.88.227.214 -
1460965304.319    164 127.0.0.1 TCP_MISS/307 1170 GET https://registry-1.docker.io/v2/sonatype/nexus/blobs/sha256:6b9b106ccb616528c20e74b5e3ed69350259f9a6985b219c576ca54e1aaa4924 - PINNED/54.88.227.214 application/octet-stream
1460965304.352    186 127.0.0.1 TCP_MISS/307 1170 GET https://registry-1.docker.io/v2/sonatype/nexus/blobs/sha256:a9945fc248582bec10c98bd760cb8374760af5681629ee92eead69e1d7240c47 - PINNED/54.88.227.214 application/octet-stream
1460965304.368    197 127.0.0.1 TCP_MISS/307 1170 GET https://registry-1.docker.io/v2/sonatype/nexus/blobs/sha256:1616b5310b71ac7ccafe70789a4c0d025928905b7f5b8c777c9ca875d4b0b8dc - PINNED/54.88.227.214 application/octet-stream
1460965304.520    200 127.0.0.1 NONE/200 0 CONNECT dseasb33srnrn.cloudfront.net:443 - HIER_DIRECT/52.85.173.34 -
1460965304.528    159 127.0.0.1 NONE/200 0 CONNECT dseasb33srnrn.cloudfront.net:443 - HIER_DIRECT/52.85.173.34 -
1460965304.534    181 127.0.0.1 NONE/200 0 CONNECT dseasb33srnrn.cloudfront.net:443 - HIER_DIRECT/52.85.173.34 -
1460965328.056  23527 127.0.0.1 TCP_MISS/200 5396563 GET https://dseasb33srnrn.cloudfront.net/registry-v2/docker/registry/v2/blobs/sha256/16/1616b5310b71ac7ccafe70789a4c0d025928905b7f5b8c777c9ca875d4b0b8dc/data? - PINNED/52.85.173.34 application/octet-stream
1460965328.712    655 127.0.0.1 NONE/200 0 CONNECT registry-1.docker.io:443 - HIER_DIRECT/54.210.75.104 -
1460965328.963    250 127.0.0.1 TCP_MISS/307 1170 GET https://registry-1.docker.io/v2/sonatype/nexus/blobs/sha256:93596c4676a6b7f1ae46d6a03681f1de30319d880dbec0a715848318981c1074 - PINNED/54.210.75.104 application/octet-stream
1460965329.281    318 127.0.0.1 NONE/200 0 CONNECT dseasb33srnrn.cloudfront.net:443 - HIER_DIRECT/52.85.173.34 -
1460965329.403    119 127.0.0.1 TCP_MISS/200 2109 GET https://dseasb33srnrn.cloudfront.net/registry-v2/docker/registry/v2/blobs/sha256/93/93596c4676a6b7f1ae46d6a03681f1de30319d880dbec0a715848318981c1074/data? - PINNED/52.85.173.34 application/octet-stream
```

**Comment by paoloantinori** [comment]
Update for those that are interested:

I have made some progress but hitting other issues. Docker uses Amazon Cloudfront as a CDN, and it does adding special query params with a signature and a time to live, that changes with every request and breaks Squid caching logic

I have tried to manipulate those params at squid level with an external filter, but that turned the urls into invalid for Cloudfront, so blocking the download of the resource.

It seems that Squid > 3.4 supports http://wiki.squid-cache.org/Features/StoreID that might help with what I'm trying to achieve, despite I'm not sure if it works at query string level.

I'm now trying to play with the newer version and see how far I can go.

**Comment by paoloantinori** [comment]
I have had a success using a local instance of a docker registry to proxy the calls to docker hub. I just need to decide with the team how to handle the life cycle of this. Ideally it has to stay decouple from the `vagrant up/down` main script

**Comment by rawlingsj** [comment]
Awesome @paoloantinori !  I'm happy if you want to test it out on me?

**Comment by paoloantinori** [comment]
Sure, I'm almost ready to publish this. But there is a limit atm:

https://github.com/docker/distribution/issues/868#issuecomment-130850532

Google docker repo supports only v1 protocol that doesn't play well with mirrors, so I cannot cache `gcr.io/google_containers/hyperkube`, so atm the only things that I'm caching are:

`{"repositories":["fabric8/fabric8-console","fabric8/jenkinshift"]}`

But I'm thinking about ways to cache other stuff.

**Comment by paoloantinori** [comment]
Done. I might need to look for some betatesters, @rawlingsj and @jstrachan 
Also because I have tested this on a Linux box with a mechanical hdd. There could be some hidden race- condition in the activation of some services. I fixed the ones I have spotted, but others could hide...

**TL;DR**

---

```
# set FABRIC8_SHARED_STORAGE to the file system location you want to use as a 
# download cache between executions, it's going to take ~800MB

time FABRIC8_SHARED_STORAGE='/500GB/test3' vagrant up

# the following is the output following a cycle of 
# vagrant destroy
# time FABRIC8_SHARED_STORAGE='/500GB/test3' vagrant up

...
real    5m42.407s
user    0m1.533s
sys 0m0.294s
```

**Description:**

---

I have enhanced the script to add 2 + 1 layers of caching:
1. Manual logic in the script to cache 3 specific Docker images
2. An instance of Squid, to act as a web cache for `curl` invocations
3. An instance of Docker Registry to act as a pass through mirror towards official Dockerhub

**Layer 0 - Manual caching at bash level :**

---

I had to add this as an entry point: I cannot cache before starting up the cache itself! :chicken: :egg:
And since I had to add this layer anyhow I'm also using it to help with a race condition, so I'm letting it also saving the Squid docker image, that I could have delegated to the registry itself.

This layer is also used to manually cache Kubernetes distro image, since that being hosted on Google private Docker registry, doesn't play well with the Docker Registry local mirror image.

Note:
Technically I could have used this approach for everything, but the other solution are transparent to their respective layers, so they might make more sense for future evolution of the project.

**Layer 1 - Web caching at `curl` level :**

---

Used to for `curl` invocations. Currently saves ~40MB of _Kubernetes_ release and _gofabric8_

**Layer 2 - Dockerhub caching at `docker` level :**

---

Used for `docker pull` invocations. Limited by the fact that works only for Dockerhub hosted images at the moment. There might be a way to make it work with `gcr.io` but I couldn't find one easily.
Currently caches `fabric8/fabric8-console` and `fabric8/jenkinshift`:

```
curl http://localhost:5000/v2/_catalog
{"repositories":["fabric8/fabric8-console","fabric8/jenkinshift"]}
```

Note: The instance of this Docker Registry Mirror works on `http`. It's possible and not too hard to deploy an `https` one but it requires the generation of self signed TLS certificates and I decided to avoid it.

**Sample view of the cached data:**

---

```
$ du -sh
728M

$ls -l
drwxr-xr-x.  3 pantinor pantinor 4.0K Apr 24 11:14 docker_registry_cache
-rw-r--r--.  1 pantinor pantinor 170M Apr 24 09:20 docker_registry.tar
drwxr-xr-x.  3 pantinor pantinor 4.0K Apr 24 11:12 hypercube
drwxrwxr-x+ 18 pantinor pantinor 4.0K Apr 24 12:41 squid_cache
-rw-r--r--.  1 pantinor pantinor  23M Apr 24 12:06 squid.tar
```
